bcm63xx: add support for Innacomm W3400V6

spliew89 opened the pull request:
Innacomm W3400V6 is an xDSL B/G wireless router
based on Broadcom BCM6328 SoC.

Hardware:
   SoC:          Broadcom BCM6328
   CPU:          BMIPS4350 V8.0, 320 MHz, 1 core
   Flash:        SPI-NOR 8MB, MX25L6406E
   RAM:          64 MB
   Ethernet:     4x 10/100 Mbps
   Switch:       Integrated
   Wireless:     802.11b/g, BCM4312
   LEDs/Buttons: 9x / 2x

Flash instruction, web UI:
1) Set a static IP on your computer compatible
with 192.168.1.1, i.e 192.168.1.100
2) Connect the ethernet cable from your computer to the router.
3) Make sure the router is powered off.
4) Press the reset button, don't release it yet!
5) While pressing reset, power on the router.
6) Wait 10 seconds or more.
Note: The power LED is red at first then turns to solid
green when ready.
8) Release the reset button.
9) Browse to 192.168.1.1
10) Select .bin file.
10) Upgrade the image.
11) Wait for it to reboot.

Note: The SPI-NOR has 4-bit Block Protection but kernel v5.4 only supports 3-bit. Therefore, some pieces of code are copied from v5.9-RC which has 4-bit support.

spliew89 committed:
bcm63xx: batch process RX path
Use netif_receive_skb_list to batch process skb in RX.
Tested on BCM6328 320 MHz using iperf3 -s 30, lowered retransmits from
500+ down to double digits.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
kernel: unlock MX25L6406E with 4 bit Block Protect
Hacked in basic support for 4 bit Block Protection register, copied from
linux-master. Needed to unlock Innacomm W3400V's SPI flash MX25L6406E,
compatible with MX25L6405D.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bcm63xx: add support for Innacomm W3400V6
Innacomm W3400V6 is an xDSL B/G wireless router
based on Broadcom BCM6328 SoC.

Hardware:
   SoC:          Broadcom BCM6328
   CPU:          BMIPS4350 V8.0, 320 MHz, 1 core
   Flash:        SPI-NOR 8MB, MX25L6406E
   RAM:          64 MB
   Ethernet:     4x 10/100 Mbps
   Switch:       Integrated
   Wireless:     802.11b/g, BCM4312
   LEDs/Buttons: 9x / 2x

Flash instruction, web UI:
1) Set a static IP on your computer compatible
with 192.168.1.1, i.e 192.168.1.100
2) Connect the ethernet cable from your computer to the router.
3) Make sure the router is powered off.
4) Press the reset button, don't release it yet!
5) While pressing reset, power on the router.
6) Wait 10 seconds or more.
Note: The power LED is red at first then turns to solid
green when ready.
8) Release the reset button.
9) Browse to 192.168.1.1
10) Select .bin file.
10) Upgrade the image.
11) Wait for it to reboot.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

adschm commented on target/linux/bcm63xx/base-files/etc/board.d/02_network:
```diff
@@ -97,7 +97,8 @@ huawei,echolife-hg556a-c)
 observa,vh4032n|\
-sagem,fast-2704n)
+sagem,fast-2704n|\
+innacomm,w3400v6)
```
Alphabetic sorting, please.

adschm commented on target/linux/bcm63xx/dts/bcm6328-innacomm-w3400v6.dts:
```diff
@@ -0,0 +1,96 @@
+/dts-v1/;
```
dts-v1 is in the dtsi now and can be dropped here.

adschm commented on target/linux/bcm63xx/dts/bcm6328-innacomm-w3400v6.dts:
```diff
@@ -0,0 +1,96 @@
+		inet_green {
+			label = "W3400V6:green:inet";
+			gpios = <&pinctrl 1 1>;
+		};
```
Please also add empty lines between nodes here.

adschm commented on target/linux/bcm63xx/dts/bcm6328-innacomm-w3400v6.dts:
```diff
@@ -0,0 +1,96 @@
+		compatible = "gpio-leds";
+
+		inet_green {
+			label = "W3400V6:green:inet";
```
Please remove the model prefix, leaving just "green:inet" etc.

adschm commented on target/linux/bcm63xx/image/bcm63xx.mk:
```diff
@@ -756,6 +756,18 @@ define Device/huawei_echolife-hg655b
+  DEVICE_MODEL := W3400V6
+  CFE_BOARD_ID := 96328ang
+  CHIP_ID := 6328
+  FLASH_MB :=8
```
Missing space.

adschm commented on target/linux/bcm63xx/patches-5.4/532-board_bcm6328-innacomm-w3400v6.patch:
```diff
@@ -0,0 +1,65 @@
+ 	{ .compatible = "sagem,fast-2704-v2", .data = &board_FAST2704V2, },
+ 	{ .compatible = "sercomm,ad1018", .data = &board_AD1018, },
+ 	{ .compatible = "sercomm,ad1018-nor", .data = &board_AD1018, },
++	{ .compatible = "innacomm,w3400v6", .data = &board_W3400V6, },
```
Alphabetic sorting, please.

spliew89 committed:
bcm63xx: w3400v6: Changes proposed by adschm.
Cosmetic changes in tune with latest upstream source code.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 commented:
Hi @adschm, I've made the changes you proposed.

spliew89 committed:
Merge branch 'master' of https://github.com/openwrt/openwrt

ynezz commented:
You need to fix your commits/rebase.

spliew89 committed:
bcm63xx: integrate board data into unified patch
Integrate board data into unified board patch to reduce number of
patches.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 commented:
> You need to fix your commits/rebase.

Sorry, I'm still a noob at this. Can you be more specific?

Noltari commented:
@spliew89 can you try to upstream the MX25L6406E patch? (http://www.linux-mtd.infradead.org/mail.html https://patchwork.ozlabs.org/project/linux-mtd/list/)
I'm not sure if this patch affects any supported devices with MX25L6405D.

adschm commented on target/linux/bcm63xx/dts/bcm6328-innacomm-w3400v6.dts:
```diff
@@ -0,0 +1,99 @@
+#include "bcm6328.dtsi"
```
Please add an SPDX license identifier.

spliew89 commented:
> 
> 
> @spliew89 can you try to upstream the MX25L6406E patch? (http://www.linux-mtd.infradead.org/mail.html https://patchwork.ozlabs.org/project/linux-mtd/list/)
> I'm not sure if this patch affects any supported devices with MX25L6405D.

The latest SPI NOR driver implementation is different from kernel v5.4's which doesn't have  4-bit BP support. Hence the hacky patch. Otherwise, supporting it looks simple and uniform across different chip models. I'll just have to upstream the MX25L6405D's feature flags change.

Existing code is already disabling 3-bit Block Protect on startup, so this patch is just adding another bit to it. It shares the same bit number as Top/Bottom Protect but that needs to be supported via a separate feature flag (which MX25L6405D doesn't), so no danger of clashing there. Overall, this patch is safe.

spliew89 committed:
bcm63xx: add SPDX license identifier to Innacomm W3400V6's DTS

Add SPDX license identifier to Innacomm W3400V6's DTS file.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bcm63xx: w3400v6: resolve conflicts

Resolve conflicts from the last rebase.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
Merge branch 'master' into a

Noltari commented:
@spliew89 can you test this branch? https://github.com/Noltari/openwrt/commits/bcm63xx-inacomm-w3400v6
It should have all the requested changes merged into 2 commits (along with some personal additional improvements).
FYI, the flash patch should be added first, since OpenWrt won't work without it.

spliew89 commented:
> 
> 
> @spliew89 can you test this branch? https://github.com/Noltari/openwrt/commits/bcm63xx-inacomm-w3400v6
> It should have all the requested changes merged into 2 commits (along with some personal additional improvements).
> FYI, the flash patch should be added first, since OpenWrt won't work without it.

I have flashed your branch into the router.
WAN-LAN works fine. WLAN is faster probably due to disabling 802.11b.
SPI NOR works fine. /etc/config files are written and retained between boots.
Everything looks OK.

```
root@OpenWrt:~# dmesg
[    0.000000] Linux version 5.4.81 (root@79abd5d83008) (gcc version 8.4.0 (Open                                                                                        Wrt GCC 8.4.0 r15160-9f150ff453)) #0 Mon Dec 7 15:37:57 2020
[    0.000000] Detected Broadcom 0x6328 CPU revision b0
[    0.000000] CPU frequency is 320 MHz
[    0.000000] 64MB of RAM installed
[    0.000000] board_bcm963xx: Boot address 0xb8000000
[    0.000000] board_bcm963xx: CFE version: 1.0.37-106.24
[    0.000000] bcm63xx_nvram: nvram checksum failed, contents may be invalid (ex                                                                                        pected 00000000, got b3df2d96)
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0002a075 (Broadcom BMIPS4350)
[    0.000000] board: board name: 96328ang
[    0.000000] MIPS: machine is Innacomm W3400V6
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 16 bytes.
[    0.000000] Primary data cache 32kB, 2-way, VIPT, cache aliases, linesize 16                                                                                         bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000]   Normal zone: 144 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: rootfstype=squashfs,jffs2 noinitrd console=t                                                                                        tyS0,115200
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes, lin                                                                                        ear)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes, line                                                                                        ar)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 57352K/65536K available (4740K kernel code, 213K rwdata,                                                                                         1076K rodata, 1260K init, 199K bss, 8184K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=16, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] random: get_random_bytes called from start_kernel+0x328/0x530 wit                                                                                        h crng_init=0
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_i                                                                                        dle_ns: 11945377789 ns
[    0.000017] sched_clock: 32 bits at 160MHz, resolution 6ns, wraps every 13421                                                                                        772796ns
[    0.008126] Calibrating delay loop... 319.74 BogoMIPS (lpj=639488)
[    0.050493] pid_max: default: 32768 minimum: 301
[    0.055631] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linea                                                                                        r)
[    0.063213] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes,                                                                                         linear)
[    0.080505] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, ma                                                                                        x_idle_ns: 7645041785100000 ns
[    0.090585] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.097803] pinctrl core: initialized pinctrl subsystem
[    0.104484] NET: Registered protocol family 16
[    0.354467] registering PCI controller with io_map_base unset
[    0.410119] PCI host bridge to bus 0000:00
[    0.414388] pci_bus 0000:00: root bus resource [mem 0x10f00000-0x10ffffff]
[    0.421452] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.428436] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.435421] pci_bus 0000:00: No busn resource found for root bus, will use [b                                                                                        us 00-ff]
[    0.443633] pci 0000:00:00.0: [14e4:6328] type 01 class 0x060400
[    0.449938] pci 0000:00:00.0: PME# supported from D0 D3hot
[    0.457060] pci 0000:00:00.0: bridge configuration invalid ([bus 00-00]), rec                                                                                        onfiguring
[    0.465666] pci 0000:01:00.0: [14e4:4312] type 00 class 0x028000
[    0.471941] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x00003fff 64bit]
[    0.478981] pci 0000:01:00.0: enabling Extended Tags
[    0.484225] pci 0000:01:00.0: supports D1 D2
[    0.490135] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.496981] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 01
[    0.503815] pci 0000:00:00.0: BAR 8: assigned [mem 0x10f00000-0x10ffffff]
[    0.510812] pci 0000:01:00.0: BAR 0: assigned [mem 0x10f00000-0x10f03fff 64bi                                                                                        t]
[    0.518344] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.523433] pci 0000:00:00.0:   bridge window [mem 0x10f00000-0x10ffffff]
[    0.530647] workqueue: max_active 576 requested for napi_workq is out of rang                                                                                        e, clamping between 1 and 512
[    0.548730] clocksource: Switched to clocksource MIPS
[    0.556420] pci 0000:00:00.0: enabling device (0000 -> 0002)
[    0.562321] b43-pci-bridge 0000:01:00.0: enabling device (0000 -> 0002)
[    0.596692] ssb: Found chip with id 0x4312, rev 0x01 and package 0x00
[    0.700944] b43-pci-bridge 0000:01:00.0: Sonics Silicon Backplane found on PC                                                                                        I device 0000:01:00.0
[    0.711041] NET: Registered protocol family 2
[    0.717531] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096                                                                                         bytes, linear)
[    0.726255] TCP established hash table entries: 1024 (order: 0, 4096 bytes, l                                                                                        inear)
[    0.734210] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.741529] TCP: Hash tables configured (established 1024 bind 1024)
[    0.748739] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.755529] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.763351] NET: Registered protocol family 1
[    0.767995] PCI: CLS 0 bytes, default 16
[    0.778539] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.802671] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.808715] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORIT                                                                                        Y) (c) 2001-2006 Red Hat, Inc.
[    0.864324] bcm6328-pinctrl 10000080.pin-controller: registered at mmio (ptrv                                                                                        al)
[    0.874538] 10000100.serial: ttyS0 at MMIO 0x10000100 (irq = 36, base_baud =                                                                                         1562500) is a bcm63xx_uart
[    0.884300] printk: console [ttyS0] enabled
[    0.892895] printk: bootconsole [early0] disabled
[    0.922956] random: fast init done
[    0.927352] spi-nor spi1.0: mx25l6405d (8192 Kbytes)
[    0.933938] bcm63xxpart: nvram checksum failed, contents may be invalid (expe                                                                                        cted 00000000, got b3df2d96)
[    0.943860] bcm63xxpart: Partition 0 is CFE offset 0 and length 10000
[    0.950478] bcm63xxpart: Partition 1 is nvram offset 7f0000 and length 10000
[    0.957727] bcm63xxpart: Partition 2 is linux offset 10000 and length 7e0000
[    0.965166] 3 bcm63xxpart partitions found on MTD device spi1.0
[    0.971277] Creating 3 MTD partitions on "spi1.0":
[    0.976210] 0x000000000000-0x000000010000 : "CFE"
[    0.983214] 0x0000007f0000-0x000000800000 : "nvram"
[    0.990422] 0x000000010000-0x0000007f0000 : "linux"
[    1.005344] parser_imagetag: rootfs: CFE image tag found at 0x0 with version                                                                                         6, board type 96328ang
[    1.014727] parser_imagetag: Partition 0 is kernel offset 100 and length 1e2b                                                                                        48
[    1.022240] parser_imagetag: Partition 1 is rootfs offset 1e2c48 and length 5                                                                                        fd3b8
[    1.030022] parser_imagetag: Spare partition is offset 480004 and length 35ff                                                                                        fc
[    1.037753] 2 bcm963xx-imagetag partitions found on MTD device linux
[    1.044319] Creating 2 MTD partitions on "linux":
[    1.049164] 0x000000000100-0x0000001e2c48 : "kernel"
[    1.056432] 0x0000001e2c48-0x0000007e0000 : "rootfs"
[    1.063641] mtd: device 4 (rootfs) set to be root filesystem
[    1.074606] 1 squashfs-split partitions found on MTD device rootfs
[    1.081069] 0x000000480000-0x0000007e0000 : "rootfs_data"
[    1.090956] libphy: Fixed MDIO Bus: probed
[    1.095896] bcm63xx_enetsw bcm63xx_enetsw.0: IRQ index 1 not found
[    1.145800] b53_common: found switch: BCM63xx, rev 0
[    1.151558] bcm63xx-wdt bcm63xx-wdt:  started, timer margin: 30 sec
[    1.160485] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    1.167720] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    1.174882] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    1.182049] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    1.189215] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    1.199347] NET: Registered protocol family 10
[    1.222089] Segment Routing with IPv6
[    1.226100] NET: Registered protocol family 17
[    1.230797] 8021q: 802.1Q VLAN Support v1.8
[    1.244457] VFS: Mounted root (squashfs filesystem) readonly on device 31:4.
[    1.269007] Freeing unused kernel memory: 1260K
[    1.273673] This architecture does not have kernel memory protection.
[    1.280279] Run /sbin/init as init process
[    2.482848] init: Console is alive
[    2.486862] init: - watchdog -
[    4.143976] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    4.441722] kmodloader: done loading kernel modules from /etc/modules-boot.d/                                                                                        *
[    4.459895] init: - preinit -
[    6.304173] random: jshn: uninitialized urandom read (4 bytes read)
[    6.391594] random: jshn: uninitialized urandom read (4 bytes read)
[    6.653134] random: jshn: uninitialized urandom read (4 bytes read)
[    7.629264] bcm63xx_enetsw bcm63xx_enetsw.0: link UP on Port 4, 100Mbps, full                                                                                        -duplex
[    7.637472] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    7.777254] urandom_read: 2 callbacks suppressed
[    7.777268] random: procd: uninitialized urandom read (4 bytes read)
[    8.636696] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   12.211879] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[   12.227465] urandom-seed: Seed file not found (/etc/urandom.seed)
[   12.444525] procd: - early -
[   12.447712] procd: - watchdog -
[   12.968785] random: crng init done
[   13.189725] procd: - watchdog -
[   13.398108] procd: - ubus -
[   13.473209] procd: - init -
[   15.449780] kmodloader: loading kernel modules from /etc/modules.d/*
[   15.528133] Loading modules backported from Linux version v5.8-0-gbcf876870b9                                                                                        5
[   15.535610] Backport generated by backports.git v5.8-1-0-g79400d9e
[   15.659926] xt_time: kernel timezone is -0000
[   16.162945] PPP generic driver version 2.4.2
[   16.180829] NET: Registered protocol family 24
[   16.335446] b43-phy0: Broadcom 4312 WLAN found (core revision 15)
[   16.384713] b43-phy0: Found PHY: Analog 6, Type 5 (LP), Revision 1
[   16.391098] b43-phy0: Found Radio: Manuf 0x17F, ID 0x2062, Revision 2, Versio                                                                                        n 0
[   16.398704] b43-phy0 warning: 5 GHz band is unsupported on this PHY
[   16.416551] urngd: v1.0.2 started.
[   16.429181] Broadcom 43xx driver loaded [ Features: PNL ]
[   16.482726] kmodloader: done loading kernel modules from /etc/modules.d/*
[   17.141124] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[  157.907531] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[  157.929292] jffs2_build_filesystem(): unlocking the mtd device...
[  157.929304] done.
[  157.937608] jffs2_build_filesystem(): erasing all blocks after the end marker                                                                                        ...
[  164.117158] bcm63xx_enetsw bcm63xx_enetsw.0: link UP on Port 3, 100Mbps, full                                                                                        -duplex
[  164.133046] bcm63xx_enetsw bcm63xx_enetsw.0: link UP on Port 4, 100Mbps, full                                                                                        -duplex
[  164.160619] br-lan: port 1(eth0.1) entered blocking state
[  164.166259] br-lan: port 1(eth0.1) entered disabled state
[  164.172488] device eth0.1 entered promiscuous mode
[  164.177440] device eth0 entered promiscuous mode
[  164.216545] br-lan: port 1(eth0.1) entered blocking state
[  164.222132] br-lan: port 1(eth0.1) entered forwarding state
[  165.117544] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[  180.026370] done.
[  180.028346] jffs2: notice: (1367) jffs2_build_xattr_subsystem: complete build                                                                                        ing xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead,                                                                                         0 orphan) found.
[  180.366312] overlayfs: upper fs does not support tmpfile.
[  410.345563] br-lan: port 1(eth0.1) entered disabled state
[  410.385653] device eth0.1 left promiscuous mode
[  410.390346] device eth0 left promiscuous mode
[  410.395152] br-lan: port 1(eth0.1) entered disabled state
[  419.261170] bcm63xx_enetsw bcm63xx_enetsw.0: link UP on Port 3, 100Mbps, full                                                                                        -duplex
[  419.269360] bcm63xx_enetsw bcm63xx_enetsw.0: link UP on Port 4, 100Mbps, full                                                                                        -duplex
[  419.287122] br-lan: port 1(eth0.1) entered blocking state
[  419.292746] br-lan: port 1(eth0.1) entered disabled state
[  419.298925] device eth0.1 entered promiscuous mode
[  419.303883] device eth0 entered promiscuous mode
[  419.332892] br-lan: port 1(eth0.1) entered blocking state
[  419.338465] br-lan: port 1(eth0.1) entered forwarding state
[  420.253568] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
```

Noltari commented:
Merged, thanks!
https://github.com/openwrt/openwrt/commit/21311ca6b6144c86bcfea28bff51eb937e34029e
https://github.com/openwrt/openwrt/commit/a58826c34f34e1ce99d0870a0a0eefd71e4868a0
